# Recommendations Using Custom Similarity Functions <sup>1</sup>

## Imports

In [1]:
import numpy as np
import pandas as pd
from scipy.special import logsumexp

## Load Data

Only loading a subset of the original data set for proof of concept reasons.

In [2]:
# 80/20 split earlier
df_train = pd.read_csv('../Data/training_data_subset.csv')
df_test = pd.read_csv('../Data/testing_data_subset.csv')

In [3]:
df_train.head(2)

category  \
0  ['Grocery & Gourmet Food', 'Candy & Chocolate'...   
1  ['Grocery & Gourmet Food', 'Jams, Jellies & Sw...   

                                               title  \
0             YumEarth Organic Gummy Bears, 10 Count   
1  Bell Plantation Powdered PB2 Bundle: 1 Peanut ...   

                                            also_buy     brand  \
0  ['B008CC8UXC', 'B00C25LO8S', 'B073RWDCMD', 'B0...  YumEarth   
1  ['B06W9N8X9H', 'B06X15V3DC', 'B01ENYJX3S', 'B0...       PB2   

                                  rank  \
0  129,438 in Grocery & Gourmet Food (   
1    1,214 in Grocery & Gourmet Food (   

                                           also_view main_cat  price  \
0  ['B008CC8UXC', 'B00C25LNWA', 'B008CC8ULY', 'B0...  Grocery    NaN   
1                                                NaN  Grocery  18.49   

         asin  overall  verified      reviewerID  vote  \
0  B008B7JNRA      3.0      True  A35KP4ROS9KWPO   NaN   
1  B00H9H56QA      5.0      True   AVAMZWS7AAI1S   NaN   

                                               style  for_testing  
0  {'Size:': ' 10 Count', 'Style:': ' Natural Gum...        False  
1            {'Size:': ' Pack of 2 (1 each flavor)'}        False

In [4]:
df_test.head(2)

category  \
0  ['Grocery & Gourmet Food', 'Snack Foods', 'Bar...   
1  ['Grocery & Gourmet Food', 'Snack Foods', 'Chi...   

                                               title  \
0                     Grocery &amp; Gourmet Food" />   
1  Gourmet Basics Smart Fries 4-Flavor Variety Pa...   

                                            also_buy           brand  \
0  ['B01MT0QDPO', 'B00NL17FE4', 'B01NBM9OJN', 'B0...   Nature Valley   
1  ['B0763SHX4W', 'B0040FIHS8', 'B00FYR5HS4', 'B0...  Gourmet Basics   

                                 rank also_view main_cat  price        asin  \
0  16,921 in Grocery & Gourmet Food (       NaN  Grocery  18.04  B001E6GFR6   
1  53,167 in Grocery & Gourmet Food (       NaN  Grocery  23.99  B003AZ2ECY   

   overall  verified      reviewerID  vote style  for_testing  
0      5.0      True  A2IUE299OONA73   NaN   NaN         True  
1      4.0      True  A38NO7J1TK4R1W   NaN   NaN         True

### RMSE

In [5]:
def compute_rmse(y_pred, y_true):
    """ Compute Root Mean Squared Error. """
    
    return np.sqrt(np.mean(np.power(y_pred - y_true, 2)))

### Evaluation method

In [6]:
def evaluate(estimate_f):
    """ RMSE-based predictive performance evaluation with pandas. """
    
    ids_to_estimate = zip(df_test.reviewerID, df_test.asin)
    estimated = np.array([estimate_f(u,i) for (u,i) in ids_to_estimate])
    real = df_test.overall.values
    return compute_rmse(estimated, real)

### Euclidean 'similarity'

$$ sim(x,y) = \frac{1}{1 + \sqrt{\sum (x - y)^2}}$$

In [7]:
def euclidean(s1, s2):
    """Take two pd.Series objects and return their euclidean 'similarity'."""
    diff = s1 - s2
    return 1 / (1 + np.sqrt(np.sum(diff ** 2)))

### Cosine similarity

$$ sim(x,y) = \frac{(x . y)}{\sqrt{(x . x) (y . y)}} $$

In [8]:
def cosine(s1, s2):
    """Take two pd.Series objects and return their cosine similarity."""
    return np.sum(s1 * s2) / np.sqrt(np.sum(s1 ** 2) * np.sum(s2 ** 2))

### Pearson correlation

$$ sim(x,y) = \frac{(x - \bar x).(y - \bar y)}{\sqrt{(x - \bar x).(x - \bar x) * (y - \bar y)(y - \bar y)}} $$

In [9]:
def pearson(s1, s2):
    """Take two pd.Series objects and return a pearson correlation."""
    s1_c = s1 - s1.mean()
    s2_c = s2 - s2.mean()
    return np.sum(s1_c * s2_c) / np.sqrt(np.sum(logsumexp(s1_c ** 2)) * np.sum(logsumexp(s2_c ** 2)))

### Jaccard similarity

$$ sim(x,y) = \frac{(x . y)}{(x . x) + (y . y) - (x . y)} $$

In [10]:
def jaccard(s1, s2):
    dotp = np.sum(s1 * s2)
    return dotp / (np.sum(s1 ** 2) + np.sum(s2 ** 2) - dotp)

def binjaccard(s1, s2):
    dotp = s1.index.intersection(s2.index).size
    return dotp / (s1.sum() + s2.sum() - dotp)

In [11]:
class CollaborativeSimilarityRecommendation:
    """ Collaborative filtering using a custom sim(u,u'). """

    def __init__(self, similarity=None):
        """ Prepare datastructures for estimation. """
        
        self.all_user_profiles = df_test.pivot_table('overall', index='asin', columns='reviewerID')
        self._similarity = similarity
        
    @property
    def similarity(self):
        return self._similarity
    
    @similarity.setter
    def similarity(self, value):
        self._similarity = value
        
    def estimate(self, user_id, product_id):
        """ Ratings weighted by custom similarity. """
        
        user_condition = df_test.reviewerID != user_id
        movie_condition = df_test.asin == product_id
        ratings_by_others = df_test.loc[user_condition & movie_condition]
        if ratings_by_others.empty: 
            return 4.0
        
        ratings_by_others.set_index('reviewerID', inplace=True)
        their_ids = ratings_by_others.index
        their_ratings = ratings_by_others.overall
        their_profiles = self.all_user_profiles[their_ids]
        user_profile = self.all_user_profiles[user_id]
        sims = their_profiles.apply(lambda profile: self.similarity(profile, user_profile), axis=0)
        ratings_sims = pd.DataFrame({'sim': sims, 'overall': their_ratings})
        ratings_sims = ratings_sims[ratings_sims.sim > 0]
        if ratings_sims.empty:
            return their_ratings.mean()
        else:
            return np.average(ratings_sims.overall, weights=ratings_sims.sim)
        

In [12]:
reco = CollaborativeSimilarityRecommendation(pearson)
print('RMSE for Pearson: %s' % evaluate(reco.estimate))

RMSE for Pearson: 1.2328263585686914


In [13]:
reco = CollaborativeSimilarityRecommendation(euclidean)
print('RMSE for Euclidean: %s' % evaluate(reco.estimate))

RMSE for Euclidean: 1.223371386810753


In [14]:
reco = CollaborativeSimilarityRecommendation(cosine)
print('RMSE for Cosine: %s' % evaluate(reco.estimate))

RMSE for Cosine: 1.2326395585090926


In [15]:
reco = CollaborativeSimilarityRecommendation(jaccard)
print('RMSE for Jaccard: %s' % evaluate(reco.estimate))

RMSE for Jaccard: 1.2270171774126628


In [16]:
reco = CollaborativeSimilarityRecommendation(binjaccard)
print('RMSE for Bin Jaccard: %s' % evaluate(reco.estimate))

RMSE for Bin Jaccard: 1.2328263585686914


## Summary
- Evaluated estimated recommendations using several custom similarity functions.
- In all cases, the estimates were:
    - very close to one another with no function being clearly better than another.
    - better than the v1 notebook.
    - worse than the v2 notebook.
- The best performing function was euclidean with a RMSE of 1.2234.
- The worse performing functions were pearson and binjaccard with a RMSE of 1.2328.

## References
1) Unata 2015 [Hands-on with PyData: How to Build a Minimal Recommendation Engine](https://www.youtube.com/watch?v=F6gWjOc1FUs).  